### ISSUE TITLE
Catch GET request to bare sso location early

### ISSUE THREAD

**Issue by phavekes** [OpenConext/OpenConext-engineblock#1537]
[This issue is imported from pivotal - Originaly created at Feb 20, 2019 by Thijs Kinkhorst](https://www.pivotaltracker.com/story/show/164121560)

Access the SSO location of Engineblock with a GET request without any parameters

(e.g.: 
https://engine.test2.surfconext.nl/authentication/idp/single-sign-on/key:test-20181127
)

Logged is:
engineblock.ERROR: Unable to find the current binding.

Highly cryptic and seems quite far in the processing chain. Check early if there\'s at least a SAMLRequest parameter present and fail early and with concrete exception if not. (Note that this endpoint supports both GET and POST).

**Comment by phavekes** [comment]
HTTP-code met indicatie dat je iets fout doet ipv dat het een server error. (Thijs Kinkhorst - Feb 26, 2019)

**Comment by phavekes** [comment]
**Some findings:**
 * EngineBlock_Corto_Module_Bindings is used to receive a SAML request (authn or response).  She uses the saml2 `Binding::getCurrentBinding` method to determine whether or not the request is valid. The saml2 `Binding::getCurrentBinding` throws the `Unable to find the current binding.` exception. 
* Our `\EngineBlock_Corto_Module_Bindings::receiveRequest` method throws a `EngineBlock_Corto_Module_Bindings_UnsupportedBindingException`, resulting in an exception message specific for ACS. This can be seen as a bug in EB right now, as `\EngineBlock_Corto_Module_Bindings::receiveRequest` is also used in SSO context, and you don\'t want a ACS error page when processing an AuthnRequest.

**How to make this more robust?**
* Option 1: By rewriting. And to be specific, by removing the `EngineBlock_Corto_Module_Bindings` and implement a modern equivelent in Symfony.
* Option 2: By adding a validation step to the `\OpenConext\EngineBlockBundle\Controller\IdentityProviderController::singleSignOnAction` where a helper class determines if the incomming AuthnRequest is supported by EB. After which it is handed over to Corto to receive the request. What will we validate? 
    * Is the correct binding used (Redirect), if not show a user friendly binding not supported exception.
    * Are the required parameter(s) present (SAMLRequest). If not, show a user friendly error message.
    *  Do we want other custom error messages? For example when the SSO endpoint is used for ACS?
* Option 3: By improving the behaviour of: `EngineBlock_Corto_Module_Bindings`

My current preference is Option 2, as Option 1 will require some more carefull planning and might quickly result in the refactoring of more than just the Bindings module.. Option 3 seems the easiest way out, but does not stimulate moving more logic to the Symfony side of EB. (Michiel Kodde - Mar 6, 2019)

**Comment by phavekes** [comment]
@thijskh, @bartgeesink, @pmeulen 
Step one of the change is finished. This entailed improving the error messages. For example. we now log the following exception when a \'bare\' GET request is performed on the SSO endpoint:

```
Mar 13 08:47:52 apps EBLOG[5845]: [2019-03-13 08:47:52] app.INFO: Handling incoming request: GET /authentication/idp/single-sign-on {"session_id":null,"request_id":"5c88b5a8763d8"} []
.............
Mar 13 08:47:52 apps EBLOG[5845]: [2019-03-13 08:47:52] engineblock.ERROR: Unsupported binding, no SAMLRequest was found in the request parameters | Caught Unhandled generic exception 
```

On to the next step: improving the error page. Have you thought about the texts we want to display in the following error situations?

1. Invalid request verb is used anyting other than a GET request will result in an `InvalidRequestMethodException` with message: `The request method "[used verb]" is not supported on the SSO endpoint`
2. Invalid binding attempted results in: `InvalidBindingException`. With message: `Unsupported binding, no SAMLRequest was found in the request parameters`. This occurs when a GET request is fired, but no valid binding could be determined by SAML2.
3. Unsupported binding was attempted: Results in `InvalidBindingException`, with message: `The binding type "SAML2\HTTPArtifact" is not supported on the SSO endpoint`.

**Please provide me with appropriate user facing error messages that can be displayed on the error feedback page.**

I suggest re-using the existing `/authentication/feedback/unable-to-receive-message` feedback page for this message. Or do you prefer the introduction of a new route? (Michiel Kodde - Mar 13, 2019)

**Comment by phavekes** [comment]
I propose the following:

"No authentication request received."

1. The request method "[used verb]" is not supported on the SSO endpoint.
1. No SAMLRequest parameter was found in the request parameters.
1. The binding type "SAML2\HTTPArtifact" is not supported on the SSO endpoint.

Please try again."

Keep it relatively technical and concrete because it is helpful for a developer to know what they\'re doing wrong..

However, do note that although we do not publish that in metadata, we do actually support POST to sso endpoint currently. So 1 and 3 would not be triggering on that scenario. But indeed on "HEAD" or artifact-binding. (Thijs Kinkhorst - Mar 13, 2019)

**Comment by phavekes** [comment]
I like @thijskh his ideas above. Some additional suggestions: 

Insert the protocol names in the messages. i.e.:
1. The HTTP request method "[used verb]" is not supported on this SAML SSO endpoint.
2. No SAMLRequest parameter was found in the HTTP request parameters.
(#2 is only for GET (HTTP-Redirect), what about the HTTP-POST binding?)
3. The urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact binding is not supported on this SAML SSO endpoint.
(People do this? Anyway, the use of the binding is easily diagnosed by the presence of the "SAMLart" HTTP GET parameter)

Finally state what is allowed / expected:
Expected a SAML 2.0 AuthnRequest using the urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect or urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST binding on this SAML SSO endpoint. (Pieter van der Meulen - Mar 13, 2019)

**Comment by phavekes** [comment]
@pmeulen Both GET and POST binding need to have a SAMLRequest parameter, right? So 2 can be for both. (Thijs Kinkhorst - Mar 13, 2019)

**Comment by phavekes** [comment]
Yes, that was a regression I was about to introduce. I was under the impression, SSO only supported HTTPRedirect binding. This is now corrected.

I now suggest for message 2:
 > \'No SAMLRequest parameter was found in the HTTP "[used verb]" request parameters\' (Michiel Kodde - Mar 13, 2019)

**Comment by phavekes** [comment]
@pmeulen Your suggestion to display the correct binding name `urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact` is not super trivial. The `SAML2\HTTPArtifact` part was merely an example and could also be SOAP. A reverse lookup from SAML2 binding to Urn would be required. This ofcourse is no big deal, but is the additional coupling to the SAML2 library worth it? (Michiel Kodde - Mar 13, 2019)

**Comment by phavekes** [comment]
@michielkodde IMO not, because I don\'t think this is common mistake to make. However, if you use your implementation #2 (do the checks before entering the saml lib, it shouldn\'t be that hard right?). Even then in nice to have territory IMO. (Pieter van der Meulen - Mar 13, 2019)

**Comment by phavekes** [comment]
@thijskh With a HTTP POST I usually call them fields (as it\'s urlencoded from data). I\'m not sure what the correct term is actually. As long as people undestand it, it\'s fine. (Pieter van der Meulen - Mar 13, 2019)

**Comment by phavekes** [comment]
(Michiel Kodde - Mar 13, 2019)

**Comment by phavekes** [comment]
@thijskh Can we get away with displaying the error message in English between the different languages? If not, I will make the exception text translatable. (Michiel Kodde - Mar 13, 2019)

**Comment by phavekes** [comment]
We can get away with it. (Thijs Kinkhorst - Mar 13, 2019)

**Comment by phavekes** [comment]
It seems to have broken unsolicited-sso-requests.

E.g. https://engine.test2.surfconext.nl/authentication/idp/unsolicited-single-sign-on/ba573f07093978e3852ddef0d2465b84/key:20181127?sp-entity-id=https://profile.test2.surfconext.nl/authentication/metadata

now triggers this error. (Thijs Kinkhorst - May 3, 2019)

**Comment by phavekes** [comment]
The request validation added for the solicited SSO was also applied on the unsolicited SSO endpoint. This caused the problem, as the unsolicited SSO request is different in many aspects. This feature was removed for now in [this PR](https://github.com/OpenConext/OpenConext-engineblock/pull/698)

I created #165821300 to add proper request validation for the unsolicited SSO endpoint. 

FYI unsolicited SSO is not covered in acceptance tests. (Michiel Kodde - May 6, 2019)
